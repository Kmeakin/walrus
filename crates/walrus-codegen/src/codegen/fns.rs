use super::*;

/// Wrapping global functions
impl<'ctx> Compiler<'ctx> {
    /// Wrap a `FunctionValue` in a closure with an empty environment
    pub fn codegen_fn_wrapper(
        &self,
        vars: &mut Vars<'ctx>,
        fn_name: &str,
        fn_value: FunctionValue<'ctx>,
        fn_type: &FnType,
    ) -> BasicValueEnum {
        let code_ptr = fn_value.as_global_value().as_pointer_value();
        let code_ptr = self.builder.build_bitcast(
            code_ptr,
            self.fn_type(vars, fn_type).ptr_type(AddressSpace::Generic),
            &format!("{fn_name}.closure"),
        );
        let closure = self
            .closure_type(vars, fn_type)
            .const_named_struct(&[code_ptr, self.codegen_null_ptr()]);
        closure.into()
    }

    /// Wrap a builtin function in a closure with an empty environment
    pub fn codegen_builtin_fn_wrapper(
        &self,
        vars: &mut Vars<'ctx>,
        builtin: Builtin,
    ) -> BasicValueEnum {
        match builtin {
            Builtin::Fn { name, ty } => {
                let fn_value = self.get_builtin_fn(vars, name, &ty);
                self.codegen_fn_wrapper(vars, name, fn_value, &ty)
            }
            Builtin::Type { .. } => {
                unreachable!("Attempt to codegen non-value builtin: {:#?}", builtin)
            }
        }
    }

    fn get_builtin_fn(
        &self,
        vars: &mut Vars<'ctx>,
        name: &str,
        fn_type: &FnType,
    ) -> FunctionValue<'ctx> {
        match vars.builtin_fns.get(name) {
            Some(fn_value) => *fn_value,
            None => {
                let fn_value = self.module.add_function(
                    &format!("builtin_{name}"),
                    self.known_fn_type(vars, fn_type),
                    None,
                );
                vars.builtin_fns.insert(name.to_string(), fn_value);
                fn_value
            }
        }
    }
}

/// Calling functions
impl<'ctx> Compiler<'ctx> {
    fn build_call(
        &'ctx self,
        fn_value: impl Into<Either<FunctionValue<'ctx>, PointerValue<'ctx>>>,
        args: &[Value<'ctx>],
        fn_name: &str,
    ) -> OptValue<'ctx> {
        let call = self
            .builder
            .build_call(fn_value.into(), args, &format!("{fn_name}.call"))
            .try_as_basic_value();
        match call {
            Left(value) => Some(value),
            Right(_) => None,
        }
    }

    pub fn codegen_fn_args(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        args: &[ExprId],
    ) -> Option<Vec<Value>> {
        args.iter()
            .map(|arg| self.codegen_expr(vars, *arg))
            .collect::<Option<Vec<_>>>()
    }

    fn codegen_known_fn_call(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        fn_id: FnDefId,
        args: &[Value<'ctx>],
    ) -> OptValue<'ctx> {
        let fn_def = &self.hir[fn_id];
        let fn_name = &self.hir[fn_def.name].as_str();
        let fn_value = vars[fn_id];
        self.build_call(fn_value, args, fn_name)
    }

    pub fn codegen_builtin_fn_call(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        name: &str,
        fn_type: &FnType,
        args: &[Value<'ctx>],
    ) -> OptValue<'ctx> {
        let fn_value = self.get_builtin_fn(vars, name, fn_type);
        self.build_call(fn_value, args, name)
    }

    fn codegen_lambda_call(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        func: ExprId,
        args: &[Value<'ctx>],
    ) -> OptValue<'ctx> {
        let mut args = args.to_vec();
        let closure_value = self.codegen_expr(vars, func)?.into_struct_value();
        let code_ptr = self
            .builder
            .build_extract_value(closure_value, 0, "closure.code")
            .unwrap()
            .into_pointer_value();
        let env_ptr = self
            .builder
            .build_extract_value(closure_value, 1, "closure.env")
            .unwrap();
        args.insert(0, env_ptr);
        self.build_call(code_ptr, &args, "lambda")
    }

    pub fn codegen_call(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        func: ExprId,
        args: &[Value<'ctx>],
    ) -> OptValue<'ctx> {
        let func_val = &self.hir[func];
        match func_val {
            Expr::Var(var_id) => {
                let var = &self.hir[*var_id];
                let denotation = self.scopes.lookup_var(*var_id, var);
                match denotation {
                    Some(Denotation::Fn(id)) => self.codegen_known_fn_call(vars, id, args),
                    Some(Denotation::Builtin(Builtin::Fn { name, ty })) => {
                        self.codegen_builtin_fn_call(vars, name, &ty, args)
                    }
                    _ => self.codegen_lambda_call(vars, func, args),
                }
            }
            _ => self.codegen_lambda_call(vars, func, args),
        }
    }
}

/// Constructing lambdas
impl<'ctx> Compiler<'ctx> {
    pub fn codegen_lambda(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        expr: ExprId,
        params: &[Param],
        body: ExprId,
    ) -> BasicValueEnum {
        let free_vars = self.free_vars(expr);

        let code_ptr = self
            .codegen_lambda_body(
                // this clone is necessary. Dont delete it!
                &mut vars.clone(),
                &free_vars,
                expr,
                params,
                body,
            )
            .as_global_value()
            .as_pointer_value();

        let closure_type = self.types[expr].as_fn().unwrap();
        let closure_alloca = self
            .builder
            .build_alloca(self.closure_type(vars, closure_type), "closure.alloca");
        let code_gep = self
            .builder
            .build_struct_gep(closure_alloca, 0, "closure.code")
            .unwrap();
        self.builder.build_store(code_gep, code_ptr);

        let env_gep = self
            .builder
            .build_struct_gep(closure_alloca, 1, "closure.env")
            .unwrap();

        // store free variables
        let free_vars_type = self.tuple_type(
            vars,
            &free_vars
                .iter()
                .map(|(pat, _)| self.types[pat].clone())
                .collect::<Vec<_>>(),
        );
        let env_alloca = self
            .builder
            .build_malloc(free_vars_type, "env.malloc")
            .unwrap();
        for (idx, (free_var, ())) in free_vars.iter().enumerate() {
            let gep = self
                .builder
                .build_struct_gep(env_alloca, idx as u32, &format!("env.{idx}.gep"))
                .unwrap();
            let val = self
                .builder
                .build_load(vars[free_var], &format!("env.{idx}"));
            self.builder.build_store(gep, val);
        }
        let env_alloca = self
            .builder
            .build_bitcast(env_alloca, self.void_ptr_type(), "env");
        self.builder.build_store(env_gep, env_alloca);
        self.builder.build_load(closure_alloca, "closure")
    }

    fn codegen_lambda_body(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        free_vars: &FreeVars,
        expr: ExprId,
        params: &[Param],
        body: ExprId,
    ) -> FunctionValue<'ctx> {
        let fn_type = self.fn_type(vars, self.types[expr].as_fn().unwrap());
        let llvm_fn = self.module.add_function("lambda", fn_type, None);
        let old_bb = self.builder.get_insert_block().unwrap();

        let bb = self.llvm.append_basic_block(llvm_fn, "lambda.entry");
        self.builder.position_at_end(bb);

        // load free vars
        let free_vars_type = self.tuple_type(
            vars,
            &free_vars
                .iter()
                .map(|(pat, _)| self.types[pat].clone())
                .collect::<Vec<_>>(),
        );

        let env_param = llvm_fn.get_nth_param(0).unwrap();
        env_param.set_name("env_ptr");
        let env_ptr = self.builder.build_bitcast(
            env_param,
            free_vars_type.ptr_type(AddressSpace::Generic),
            "env_ptr",
        );
        let env = self.builder.build_load(env_ptr.into_pointer_value(), "env");
        for (idx, (free_var, ())) in free_vars.iter().enumerate() {
            let val = self
                .builder
                .build_extract_value(env.into_struct_value(), idx as u32, &format!("env.{idx}"))
                .unwrap();
            self.codegen_local_var(vars, free_var, val)
        }

        // load params
        llvm_fn
            .get_param_iter()
            .skip(1)
            .zip(params)
            .enumerate()
            .for_each(|(idx, (llvm_param, hir_param))| {
                llvm_param.set_name(&format!("params.{idx}"));
                self.codegen_pat(vars, hir_param.pat, llvm_param)
            });

        if let Some(value) = self.codegen_expr(vars, body) {
            self.builder.build_return(Some(&value));
        }

        self.builder.position_at_end(old_bb);

        llvm_fn
    }
}
