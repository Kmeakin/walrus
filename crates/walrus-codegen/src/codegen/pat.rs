use super::*;

impl<'ctx> Compiler<'ctx> {
    pub fn codegen_pat(&self, vars: &mut Vars<'ctx>, pat_id: PatId, value: BasicValueEnum<'ctx>) {
        let pat = &self.hir[pat_id];
        match pat {
            Pat::Lit(_) | hir::Pat::Ignore => {}
            hir::Pat::Var { var, .. } => self.codegen_local_var(vars, *var, value),
            hir::Pat::Tuple(pats) => pats.iter().enumerate().for_each(|(idx, id)| {
                let val = self.get_tuple_field(value, idx);
                self.codegen_pat(vars, *id, val)
            }),
            hir::Pat::Struct { fields, .. } => {
                let struct_id = self.types[pat_id].as_struct().unwrap();
                for field in fields {
                    let val = self.get_struct_field(struct_id, value, field.name);
                    match field.pat {
                        None => self.codegen_local_var(vars, field.name, val),
                        Some(pat) => self.codegen_pat(vars, pat, val),
                    }
                }
            }
            Pat::Enum {
                variant, fields, ..
            } => {
                let enum_id = self.types[pat_id].as_enum().unwrap();
                let (payload, variant) = self.get_enum_payload(enum_id, *variant, value);
                fields.iter().for_each(|field| {
                    let val = self.get_variant_field(enum_id, &variant, payload, field.name);
                    match field.pat {
                        None => self.codegen_local_var(vars, field.name, val),
                        Some(pat) => self.codegen_pat(vars, pat, val),
                    }
                })
            }
        }
    }

    pub fn codegen_match_attempt(
        &self,
        case_idx: usize,
        test: BasicValueEnum<'ctx>,
        pat_id: PatId,
    ) -> IntValue<'ctx> {
        let pat = &self.hir[pat_id];

        match pat {
            Pat::Lit(lit) => match lit {
                Lit::Bool(b) => self.builder.build_int_compare(
                    IntPredicate::EQ,
                    test.into_int_value(),
                    self.codegen_bool(*b),
                    "Bool.eq",
                ),
                Lit::Int(i) => self.builder.build_int_compare(
                    IntPredicate::EQ,
                    test.into_int_value(),
                    self.codegen_int(*i),
                    "Int.eq",
                ),
                Lit::Char(c) => self.builder.build_int_compare(
                    IntPredicate::EQ,
                    test.into_int_value(),
                    self.codegen_char(*c),
                    "Char.eq",
                ),
                Lit::Float(f) => self.builder.build_float_compare(
                    FloatPredicate::OEQ,
                    test.into_float_value(),
                    self.codegen_float(*f),
                    "Float.eq",
                ),
                Lit::String(s) => todo!(),
            },
            Pat::Var { .. } | Pat::Ignore => self.codegen_true(),
            Pat::Tuple(pats) => self.codegen_all(pats.iter().enumerate().map(|(idx, pat)| {
                let val = self.get_tuple_field(test, idx);
                self.codegen_match_attempt(idx, val, *pat)
            })),
            Pat::Struct { fields, .. } => {
                let struct_id = self.types[pat_id].as_struct().unwrap();
                self.codegen_all(fields.iter().map(|field| match field.pat {
                    None => self.codegen_true(),
                    Some(pat) => {
                        let val = self.get_struct_field(struct_id, test, field.name);
                        self.codegen_match_attempt(case_idx, val, pat)
                    }
                }))
            }
            Pat::Enum {
                variant, fields, ..
            } => {
                let enum_id = self.types[pat_id].as_enum().unwrap();
                let enum_def = &self.hir[enum_id];
                let enum_name = &self.hir[enum_def.name];

                let (variant_idx, enum_variant) =
                    enum_def.get_variant(&self.hir, *variant).unwrap();
                let variant_name = &self.hir[enum_variant.name];

                let discriminant_matched = match self.enum_discriminant_type(enum_id) {
                    None => self.codegen_true(),
                    Some(int_type) => {
                        let discriminant = self.get_enum_discriminant(enum_id, test);
                        self.builder.build_int_compare(
                            IntPredicate::EQ,
                            discriminant,
                            int_type.const_int(variant_idx as u64, false),
                            &format!("{enum_name}::{variant_name}.cmp_discriminant"),
                        )
                    }
                };
                let bb = self.builder.get_insert_block().unwrap();
                let end_bb = self.llvm.insert_basic_block_after(
                    bb,
                    &format!("match.case{case_idx}.{enum_name}::{variant_name}.end"),
                );
                let then_bb = self.llvm.insert_basic_block_after(
                    bb,
                    &format!("match.case{case_idx}.{enum_name}::{variant_name}.then"),
                );

                self.builder
                    .build_conditional_branch(discriminant_matched, then_bb, end_bb);

                // then branch
                self.builder.position_at_end(then_bb);
                let (payload, variant) = self.get_enum_payload(enum_id, *variant, test);
                let then_value = self.codegen_all(fields.iter().map(|field| match field.pat {
                    None => self.codegen_true(),
                    Some(pat) => {
                        let val = self.get_variant_field(enum_id, &variant, payload, field.name);
                        self.codegen_match_attempt(case_idx, val, pat)
                    }
                }));
                self.builder.build_unconditional_branch(end_bb);

                // merge the 2 branches
                self.builder.position_at_end(end_bb);
                let phi = self.builder.build_phi(
                    self.llvm.bool_type(),
                    &format!("match.case{case_idx}.{enum_name}::{variant_name}.phi"),
                );
                phi.add_incoming(&[(&then_value, then_bb), (&self.codegen_false(), bb)]);
                phi.as_basic_value().into_int_value()
            }
        }
    }

    fn codegen_all(&self, bools: impl Iterator<Item = IntValue<'ctx>>) -> IntValue<'ctx> {
        bools.fold(self.codegen_true(), |acc, e| {
            self.builder.build_and(acc, e, "")
        })
    }
}
