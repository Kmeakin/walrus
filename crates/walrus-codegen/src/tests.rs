use crate::codegen::Context;
use inkwell::{memory_buffer::MemoryBuffer, OptimizationLevel};
use insta::*;

#[track_caller]
fn test_codegen(src: &str) { test::<()>(src, None) }

#[track_caller]
fn test_codegen_and_run<T>(src: &str, expected: T)
where
    T: PartialEq + std::fmt::Debug,
{
    test(src, Some(expected))
}

fn test<T>(src: &str, expected: Option<T>)
where
    T: PartialEq + std::fmt::Debug,
{
    let syntax = walrus_parser::parse(src);
    let hir = walrus_semantics::hir::lower(&syntax);
    let scopes = walrus_semantics::scopes::scopes(&hir);
    let types = walrus_semantics::ty::infer(hir.clone(), scopes.clone());

    dbg!(&syntax);
    dbg!(&hir);
    dbg!(&scopes);
    dbg!(&types);

    let module = crate::codegen("module", hir.hir, scopes, types);

    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("../snapshots");
    settings.set_prepend_module_to_snapshot(false);
    settings.bind(|| assert_display_snapshot!(module));

    let llvm = Context::create();
    let module_buffer = MemoryBuffer::create_from_memory_range_copy(module.as_bytes(), "module");
    let llvm_module = llvm.create_module_from_ir(module_buffer).unwrap();

    if let Some(expected) = expected {
        let exec_engine = llvm_module
            .create_jit_execution_engine(OptimizationLevel::None)
            .unwrap();
        let f = unsafe { exec_engine.get_function::<unsafe extern "C" fn() -> T>("main") }.unwrap();
        assert_eq!(unsafe { f.call() }, expected);
    }
}

mod literals {
    use super::*;
    #[test]
    fn empty_fn() { test_codegen_and_run("fn main() -> _ {}", ()) }

    #[test]
    fn lit_true() { test_codegen_and_run(r#"fn main() -> _ { true }"#, true) }

    #[test]
    fn lit_false() { test_codegen_and_run(r#"fn main() -> _ { false }"#, false) }

    #[test]
    fn lit_int() { test_codegen_and_run(r#"fn main() -> _ { 1 }"#, 1_i32) }

    #[test]
    fn lit_float() { test_codegen_and_run(r#"fn main() -> _ { 1.234 }"#, 1.234_f32) }

    #[test]
    fn lit_char() { test_codegen_and_run(r#"fn main() -> _ { 'a' }"#, 'a') }

    #[test]
    fn lit_string() {
        test_codegen_and_run(r#"fn main() -> _ { "hello world" }"#, "hello world".len())
    }
}

mod tuples {
    use super::*;
    #[test]
    fn tuple0() { test_codegen_and_run(r#"fn main() -> _ { () }"#, ()) }

    #[test]
    fn tuple1() { test_codegen_and_run(r#"fn main() -> _ { (1,) }"#, (1_i32,)) }

    #[test]
    fn tuple2() { test_codegen_and_run(r#"fn main() -> _ { (1, 2) }"#, (1_i32, 2_i64)) }
}

mod if_ {
    use super::*;
    #[test]
    fn if_then_else_true() {
        test_codegen_and_run("fn main() -> _ { if true {1} else {0} }", 1_i32)
    }

    #[test]
    fn if_then_else_false() {
        test_codegen_and_run("fn main() -> _ { if false {1} else {0} }", 0_i32)
    }

    #[test]
    fn if_then_true() { test_codegen_and_run(r#"fn main() -> _ { if true {1} }"#, ()); }
    #[test]
    fn if_then_false() { test_codegen_and_run(r#"fn main() -> _ { if false {1} }"#, ()); }
}

mod let_ {
    use super::*;

    #[test]
    fn let_var() { test_codegen_and_run(r#"fn main() -> _ { let x = 5; x }"#, 5_i32); }

    #[test]
    fn let_tuple2() { test_codegen_and_run(r#"fn main() -> _ { let (x, y) = (5, 6); x }"#, 5_i32) }

    #[test]
    fn let_tuple4() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    let (a, b, c, d) = (9, 8, 7, 6);
    let (x, (y, z), w) = (d, (c, b), a);
    x
}"#,
            6_i32,
        )
    }
}

mod fns {
    use super::*;
    #[test]
    fn id_fn() {
        test_codegen_and_run(
            r#"
fn main() -> _ { id(5) }
fn id(x: _) -> _ {x}
"#,
            5_i32,
        )
    }

    #[test]
    fn get_five_fn() {
        test_codegen_and_run(
            r#"
fn main() -> _ { get_five() }
fn get_five() -> _ {5}
"#,
            5_i64,
        )
    }

    #[test]
    fn fn_value() {
        test_codegen_and_run(
            r#"
fn main() -> _ { 
    let f = get_five;
    f()
}

fn get_five() -> _ {5}
"#,
            5_i32,
        )
    }

    #[test]
    fn builtin_fn_value() {
        test_codegen_and_run(
            r#"
fn main() -> _ { 
    let f = putchar;
    f('a')
}
"#,
            (),
        )
    }
}

mod return_ {
    use super::*;
    #[test]
    fn early_return() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    return 5;
    6
}
"#,
            5_i64,
        )
    }

    #[test]
    fn multiple_return() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    return 5;
    return 6;
}
"#,
            5_i64,
        )
    }

    #[test]
    fn multi_expr_return() { test_codegen_and_run(r#"fn main() -> _ { 1 + (return 2) }"#, 2); }

    #[test]
    #[ignore]
    fn if_then_return() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    if true {return 1}
    else {0}
}
"#,
            1,
        )
    }
}

mod bool_ops {
    use super::*;
    #[test]
    fn bool_not() { test_codegen_and_run(r#"fn main() -> _  {!false}"#, true) }
    #[test]
    fn bool_or() { test_codegen_and_run(r#"fn main() -> _  {false || true}"#, true) }

    #[test]
    fn bool_and() { test_codegen_and_run(r#"fn main() -> _ {true && false}"#, false) }
}

mod int_ops {
    use super::*;

    #[test]
    fn int_neg() { test_codegen_and_run(r#"fn main() ->   _  {-1}"#, -1_i32) }
    #[test]
    fn int_plus() { test_codegen_and_run(r#"fn main() ->   _  {+1}"#, 1_i32) }

    #[test]
    fn int_add() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x + y}"#, 2_i32) }
    #[test]
    fn int_sub() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x - y}"#, 0_i32) }
    #[test]
    fn int_mul() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 2; x * y}"#, 2_i32) }
    #[test]
    fn int_div() { test_codegen_and_run(r#"fn main() -> _ {let x = 5; let y = 2; x / y}"#, 2_i32) }

    #[test]
    fn int_eq() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x == y}"#, true) }
    #[test]
    fn int_neq() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x != y}"#, false) }
    #[test]
    fn int_less() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x < y}"#, false) }
    #[test]
    fn int_less_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x <= y}"#, true)
    }
    #[test]
    fn int_greater() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x > y}"#, false)
    }
    #[test]
    fn int_greater_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x >= y}"#, true)
    }
}

mod float_ops {
    use super::*;

    #[test]
    fn float_neg() { test_codegen_and_run(r#"fn main() -> _  {-1.0}"#, -1.0_f32) }
    #[test]
    fn float_plus() { test_codegen_and_run(r#"fn main() -> _  {+1.0}"#, 1.0_f32) }

    #[test]
    fn float_add() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 1.0; x + y}"#,
            2.0_f32,
        )
    }
    #[test]
    fn float_sub() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 1.0; x - y}"#,
            0.0_f32,
        )
    }
    #[test]
    fn float_mul() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 2.0; x * y}"#,
            2.0_f32,
        )
    }
    #[test]
    fn float_div() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 5.0; let y = 2.0; x / y}"#,
            2.5_f32,
        )
    }

    #[test]
    fn float_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x == y}"#, true)
    }
    #[test]
    fn float_neq() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 1.0; x != y}"#,
            false,
        )
    }
    #[test]
    fn float_less() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x < y}"#, false)
    }
    #[test]
    fn float_less_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x <= y}"#, true)
    }
    #[test]
    fn float_greater() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x > y}"#, false)
    }
    #[test]
    fn float_greater_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x >= y}"#, true)
    }
}

mod assign {
    use super::*;

    #[test]
    fn assign_var() {
        test_codegen_and_run(
            r#"fn main() -> _ {
        let mut x = 5;
        x = 6;
        x
    }"#,
            6_i32,
        )
    }

    #[test]
    fn assign_field() {
        test_codegen_and_run(
            r#"
            struct S {x: Int}

            fn main() -> _ {
                let mut s = S {x: 0};
                s.x = 1;
                s.x
        }"#,
            1_i32,
        )
    }
}

mod builtins {
    use super::*;

    #[test]
    fn builtin_exit() { test_codegen("fn main() -> Int { exit(1) }") }

    #[test]
    fn builtin_putchar() {
        test_codegen_and_run(
            r#"
fn main() -> () {
    putchar('h');
    putchar('e');
    putchar('l');
    putchar('l');
    putchar('o');
    putchar(' ');
    putchar('w');
    putchar('o');
    putchar('r');
    putchar('l');
    putchar('d');
}
"#,
            (),
        )
    }
}

mod loops {
    use super::*;

    // this successfully loops forever, which makes the test never complete!
    #[test]
    fn loop_forever() { test_codegen("fn main() -> Int { loop {} }") }

    // this successfully loops forever, which makes the test never complete!
    #[test]
    fn loop_and_continue() { test_codegen("fn main() -> () { loop {let x = 5; continue} }") }

    #[test]
    fn loop_and_break() {
        test_codegen_and_run("fn main() -> _ { loop {let x = 5; break x} }", 5_i32)
    }

    #[test]
    #[ignore]
    fn loop_and_break_and_continue() {
        test_codegen_and_run(
            "fn main() -> _ { loop { if true {continue} else {break} } }",
            (),
        )
    }
}

mod structs {
    use super::*;
    #[test]
    fn struct_constructor() {
        test_codegen_and_run(
            r#"
struct Foo {x: Int}
fn main() -> _ {
    let foo = Foo {x:5};
    foo
}
"#,
            5_i32,
        )
    }

    #[test]
    fn field_expr() {
        test_codegen_and_run(
            r#"
struct Foo {x: Int,y:Int}
fn main() -> _ {
    let foo = Foo {y:6, x:5};
    foo.y
}
"#,
            6_i32,
        )
    }

    #[test]
    fn struct_pat() {
        test_codegen_and_run(
            r#"
struct Foo {x: Int, y: Int}
fn main() -> _ {
    let Foo {y,x} = Foo{x: 5, y: 10};
    y - x
}
    "#,
            5_i32,
        );
    }

    #[test]
    fn struct_in_struct() {
        test_codegen_and_run(
            r#"
struct Foo {bar: Bar}
struct Bar {x: Int, y: Int}
fn main() -> _ {
    let bar = Bar{x: 5, y: 10};
    let foo = Foo{bar: bar};
    foo.bar.y - foo.bar.x
}
    "#,
            5_i32,
        );
    }
}

mod enums {
    use super::*;

    #[test]
    fn enum0() {
        test_codegen(
            r#"
        enum Void {}
        fn main() -> _ {
            let impossible: Void = exit(0);
            impossible
        }
    "#,
        )
    }

    #[test]
    fn enum1() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {} 
        }
        fn main() -> _ {
            Foo::Bar{}
        }
    "#,
            ((), ()),
        )
    }

    #[test]
    fn enum1_with_payload() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {x: Int} 
        }
        fn main() -> _ {
            Foo::Bar{x: 5}
        }
    "#,
            (5_i32, ()),
        )
    }

    #[test]
    fn enum2_variant0() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {},
            Baz {}
        }
        fn main() -> _ {
            Foo::Bar{}
        }
    "#,
            (0_u8, ()),
        );
    }
    #[test]
    fn enum2_variant1() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {},
            Baz {}
        }
        fn main() -> _ {
            Foo::Baz{}
        }
    "#,
            (1_u8, ()),
        )
    }

    #[test]
    fn enum2_with_payload_variant0() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {x: Int},
            Baz {y: Bool} 
        }
        fn main() -> _ {
            Foo::Bar{x: 5}
        }
    "#,
            (0_u8, 5_i64),
        )
    }

    #[test]
    #[ignore]
    fn enum2_with_payload_variant1() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {x: Int}, 
            Baz {y: Bool} 
        }
        fn main() -> _ {
            Foo::Baz{y: true}
        }
    "#,
            (1_u8, 1_i32),
        )
    }
}

mod pattern_matching {
    use super::*;

    #[test]
    fn simple_pattern_match() {
        test_codegen_and_run(
            r#"
enum Option {
    None {},
    Some {x: Int},
}

fn main() -> _ {
    let opt = Option::Some{x:1};
    match opt {
        Option::None{} => 0,
        Option::Some{x} => x,
    }
}
    "#,
            1_i32,
        )
    }

    #[test]
    fn add_options() {
        test_codegen_and_run(
            r#"
enum Option {
    None {},
    Some {x: Int},
}

fn add_options(a: Option, b: Option) -> Option {
    match (a, b) {
        (Option::Some{x:a}, Option::Some{x:b}) => Option::Some{x: a + b},
        _ => Option::None{},
    }
}

fn main() -> _ {
    let a = Option::Some{x:1};
    let b = Option::Some{x:2};
    add_options(a,b)
}
    "#,
            (1_i8, 3_i64),
        )
    }

    #[test]
    fn list_len() {
        test_codegen_and_run(
            r#"
enum List {
    Nil {},
    Cons {head: Int, tail: List}
}

fn length(l: List) -> Int {
    match l {
        List::Nil {} => 0,
        List::Cons{head: _, tail} => 1 + length(tail),
    }
}

fn main() -> _ {
    let l = List::Cons{head: 2, tail: List::Cons {head: 1, tail: List::Nil{}}};
    length(l)
}
        "#,
            2_i32,
        )
    }

    #[test]
    fn list_map() {
        test_codegen_and_run(
            r#"
enum List {
    Nil {},
    Cons {head: Int, tail: List}
}
 
fn map(l: List, f: (Int) -> Int) -> List {
    match l {
        List::Nil {} => List::Nil{},
        List::Cons{head, tail} => List::Cons{head: f(head), tail: map(tail, f)}
    }
}

fn sum(l: List) -> Int {
    match l {
        List::Nil {} => 0,
        List::Cons{head, tail} => head + sum(tail),
    }
}

fn main() -> _ {
    let l1 = List::Cons{head: 4, tail: List::Cons{head: 3, tail: List::Cons{head: 2, tail: List::Cons{head: 1, tail: List::Nil {}}}}};
    let l2 = map(l1, (x) => x * x);
    let s = sum(l2);
    s
}
        "#,
            (4 * 4 + 3 * 3 + 2 * 2 + 1) as i32,
        )
    }

    #[test]
    fn danglin_ptr() {
        test_codegen_and_run(
            r#"
enum List {
    Nil {},
    Cons {head: Int, tail: List}
}
 
fn map(l: List, f: (Int) -> Int) -> List {
    match l {
        List::Nil {} => List::Nil{},
        List::Cons{head, tail} => List::Cons{head: f(head), tail: map(tail, f)}
    }
}

fn sum(l: List) -> Int {
    match l {
        List::Nil {} => 0,
        List::Cons{head, tail} => head + sum(tail),
    }
}

fn get_list() -> _ {
    let l1 = List::Cons{head: 4, tail: List::Cons{head: 3, tail: List::Cons{head: 2, tail: List::Cons{head: 1, tail: List::Nil {}}}}};
    l1
}

fn main() -> _ {
    let l1 = get_list();
    let l2 = map(l1, (x) => x * x);
    let s = sum(l2);
    s
}
        "#,
            (4 * 4 + 3 * 3 + 2 * 2 + 1) as i32,
        )
    }
}
