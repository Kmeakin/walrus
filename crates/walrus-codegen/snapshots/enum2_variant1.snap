---
source: crates/walrus-codegen/src/tests.rs
expression: module

---
; ModuleID = 'module'
source_filename = "module"

%Foo = type { i8, {} }

define %Foo @main() {
main.entry:
  %Foo.alloca = alloca %Foo, align 8
  %Foo.discriminant.gep = getelementptr inbounds %Foo, %Foo* %Foo.alloca, i32 0, i32 0
  store i8 1, i8* %Foo.discriminant.gep, align 1
  %Foo.payload.gep = getelementptr inbounds %Foo, %Foo* %Foo.alloca, i32 0, i32 1
  %Foo.load = load %Foo, %Foo* %Foo.alloca, align 1
  ret %Foo %Foo.load
}

